File: 0543_diameter_of_binary_tree.kt
// O(n) time and space complexity. Uses a DFS recursive function
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    var maxDiameter = 0

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        dfsTravel(root)

        return maxDiameter
    }

    fun dfsTravel(node: TreeNode?): Int {
        if (node == null) return 0

        val leftHeight = dfsTravel(node.left)
        val rightHeight = dfsTravel(node.right)
        val currentDiameter = leftHeight + rightHeight // Trick: The diameter of a node in a binary tree is the sum of their left and right height.

        maxDiameter = max(maxDiameter, currentDiameter)
        return 1 + max(leftHeight, rightHeight)
    }
}

File: 0543_diameter_of_binary_tree.cpp
// https://leetcode.com/problems/diameter-of-binary-tree/
// O(n) space and time complexity solution using a DFS search algorithm
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        dfsTravel(root);

        return currentMaxDiameter;
    }
private: 
    int currentMaxDiameter = 0;

    // @returns: The height of the tree
    int dfsTravel(TreeNode* node) {
        if (node == nullptr) return 0;

        int leftHeight = dfsTravel(node->left);
        int rightHeight = dfsTravel(node->right);

        int diameterOfTree = leftHeight + rightHeight; // Diameter of a tree = height of the left and right subtrees

        if (diameterOfTree > currentMaxDiameter) {
            currentMaxDiameter = diameterOfTree;
        }

        return (1 + max(leftHeight, rightHeight));
    }
};
